src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp

```diff
@@ -93,46 +128,77 @@ void Bank::evaluate_and_allocate() {
         }
         storage_guard.unlock();
 
+        std::map<int64_t, ov::Tensor> uids_to_allocated;
+        std::map<int64_t, LazyTensor::Meta> uids_to_meta;
+
         ov::parallel_for(vec.size(), [&](std::size_t idx) {
             const auto& lt = vec[idx];
             std::unique_lock dev_guard(device_bank.mutex);
             auto iter_device_registered = device_bank.registered_tensors.find(lt);
             NPUW_ASSERT(iter_device_registered != device_bank.registered_tensors.end() &&
                         "Tensor should be registered first!");
-            if (device_bank.storage[iter_device_registered->second].tensor) {
+            auto uid = iter_device_registered->second;
+            if (device_bank.storage[uid].tensor) {
                 // Already allocated
                 return;
             }
             dev_guard.unlock();
 
             // Allocation and/or evaluation needed
-            // Evaluate concurrently, lock the device
-            // mutex only to update the device bank (& allocate on-device memory, if needed)
-            const auto& transformed_tensor = lt.eval();
+            const auto& transformed_tensor_meta = lt.eval_meta();
 
             std::unique_lock<std::mutex> guard(device_bank.mutex);
+            uids_to_meta[uid] = transformed_tensor_meta;
             if (device_for_alloc == "CPU") {
                 // No allocation needed
-                device_bank.storage[device_bank.registered_tensors.at(lt)].tensor = transformed_tensor;
+                guard.unlock();
+                auto transformed = lt.eval();
+                std::unique_lock guard(device_bank.mutex);
+                device_bank.storage.at(uid).tensor = std::move(transformed);
                 return;
             }
 
+            // Need to allocate
+            // Note: secured by the lock above
+            uids_to_allocated.insert({uid, ov::Tensor()});
+        });
+
+        std::unique_lock guard(device_bank.mutex);
+        // Allocate memory sequentially - in order of UID
+        for (const auto& elem : uids_to_allocated) {
+            auto uid = elem.first;
+
```

Comment:
This is really a vector. Instead of empty tensors, you could record meta for tensors to allocate there